### PR TITLE
Clang format

### PR DESCRIPTION
**Pull request by miguelteixeiraa**
ref: https://github.com/ada-url/ada/issues/170

I was thinking about the following options:
- Making `run-clang-format.sh`  a "cmake script" (to be used like `cmake -P clang-format.cmake`) 
- Having a Makefile in the root of the repository to add scripts like this
- Leave it as it is right there in the tools folder

Any thoughts?

### CODE REVIEW

File: .clang-format
Comment (anonrig) [inline .clang-format:2]: Can't we just use basedonstyle: google?

Ex: https://github.com/googleapis/google-cloud-cpp/blob/main/.clang-format

Comment (miguelteixeiraa) [reply .clang-format:2]: Sure thing! This is definitely better, didn't know it was possible.

**Comment by anonrig** [comment]
I'd go with something like `execute_process` that runs on every build step. Depending on the environment variable of github workflow, you can throw an error on CI builds to say that "linting is required".

If you want to run directly the format command using cmake, make as `add_custom_command()` function.

**Comment by anonrig** [comment]
Hi @miguelteixeiraa, would you like to complete this? We can open another pull request to fix all the linter related issues.

**Comment by miguelteixeiraa** [comment]
I could complete that, but if you already have everything ready, no problem! You can push what you have and we can close this one! : )

**Comment by anonrig** [comment]
> I could complete that, but if you already have everything ready, no problem! You can push what you have and we can close this one! : )

I haven't worked on it, I'm just interested :-)

**Comment by miguelteixeiraa** [comment]
I made the execution of the "check" function/script be conditioned to a variable called IS_CI, and the default "format" script/function for every time it is not the CI.. if I turn it to be a custom command, it would be better (talking about the "format")?

File: cmake/clang-format.cmake
Comment (anonrig) [inline cmake/clang-format.cmake:12]: Can you also update one of the github workflows to add a IS_CI environment variable?

Comment (anonrig) [inline cmake/clang-format.cmake:12]: We don't need to set `IS_CI` cmake variable, since it's not used anywhere else.

```suggestion
  if(DEFINED ENV{IS_CI})
```

**Review by anonrig — APPROVED** [review]
I'm happy with this pull request! Just updating or adding aa github workflow would be sufficient to merge

**Review by anonrig — APPROVED** [review]

**Comment by anonrig** [comment]
Can you remove the fixes to all files? It will create huge conflict with the aggregator pull request. Lets merge it with a failing test

File: tools/lint_and_format.py
Comment (miguelteixeiraa) [reply tools/lint_and_format.py:29]: Yeah, makes sense!

File: .github/workflows/lint_and_format_check.yml
Comment (miguelteixeiraa) [reply .github/workflows/lint_and_format_check.yml:16]: Yeah, I don't see why not!

**Comment by miguelteixeiraa** [comment]
> Can you remove the fixes to all files? It will create huge conflict with the aggregator pull request. Lets merge it with a failing test

Sure thing! I'm just trying to figure out why it still falling even formatting all files. I'll rollback after that.

**Comment by miguelteixeiraa** [comment]
It doesn't seem to be a big problem.. There is a major version difference between the version installed by default in ubuntu-latest and the one I have running locally (15.0.6, and 14.x.x on Ubuntu-latest probably). So this is formatting differently in the two places. 
**Ie we need to choose a version of clang-format to use.**

**Comment by miguelteixeiraa** [comment]
- I fixed a clang-format version (15) and the github actions workflow  is already modified to consider this version.
- Formatting will only happen automatically on each build if the FORMAT_ENABLED environment variable is set.
- The check (without formatting) will only happen if the environment variable LINT_AND_FORMAT_CHECK is set (it is set/fixed in the CI/lint-and-format workflow).
- If the script runs (both for checking or for formatting) and the executable clang-format installed has a different version than the script's default (or the one passed as an argument for the script), the script should produce an error.

That's how everything is working now : )

**Comment by miguelteixeiraa** [comment]
Now I'll go back to work on one of the tests PR  because they are probably celebrating their birthday at this point :sweat_smile:

**Comment by miguelteixeiraa** [comment]
@lemire @anonrig Do you need me to do something else on this one before merging?

**Review by anonrig — APPROVED** [review]

**Comment by lemire** [comment]
I leave it to you.

**Comment by anonrig** [comment]
@miguelteixeiraa Can we update idna with these changes too?

**Comment by miguelteixeiraa** [comment]
> @miguelteixeiraa Can we update idna with these changes too?

Sure thing!

**Pull request merged**
